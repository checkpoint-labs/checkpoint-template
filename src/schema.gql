scalar Text

type Post {
  id: String!
  author: String!
  content: Text!
  tag: String
  tx_hash: String!
  created_at: Int!
  created_at_block: Int!
}
