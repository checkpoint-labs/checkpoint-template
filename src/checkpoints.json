[
  {
    "contract": "0x04d10712e72b971262f5df09506bbdbdd7f729724030fa909e8c8e7ac2fd0012",
    "blocks": [185778, 185780, 220016, 220138, 221984]
  }
]
