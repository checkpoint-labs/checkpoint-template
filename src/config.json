{
  "network_node_url": "https://rpc.brovider.xyz/5",
  "sources": [
    {
      "contract": "0x000000000000cd17345801aa8147b8D3950260FF",
      "start": 6102403,
      "abi": "Poster",
      "events": [
        {
          "name": "NewPost(address,string,string)",
          "fn": "handleNewPost"
        }
      ]
    }
  ]
}
