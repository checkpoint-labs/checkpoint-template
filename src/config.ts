import { CheckpointConfig } from '@snapshot-labs/checkpoint';
import Poster from './abis/Poster.json';

const CONFIG = {
  mainnet: {
    networkNodeUrl: 'https://starknet-mainnet.infura.io/v3/c82b1cf516984b599108487a1b6481c4',
    contract: '0x0654e9232d5f402829755029901f69c32b423ded0f8c081e416e3b24f5a7a46e',
    start: 639485
  },
  sepolia: {
    networkNodeUrl: 'https://starknet-sepolia.infura.io/v3/c82b1cf516984b599108487a1b6481c4',
    contract: '0x03aa7630a4f9c5108bf3cd1910c7d45404cba865fc0fc0756bf9eedc073a98a9',
    start: 65137
  }
};

export function createConfig(indexerName: keyof typeof CONFIG): CheckpointConfig {
  const { networkNodeUrl, contract, start } = CONFIG[indexerName];

  return {
    network_node_url: networkNodeUrl,
    optimistic_indexing: false,
    fetch_interval: 15000,
    sources: [
      {
        contract,
        start,
        abi: 'Poster',
        events: [
          {
            name: 'NewPost',
            fn: 'handleNewPost'
          }
        ]
      }
    ],
    abis: {
      Poster
    }
  };
}
