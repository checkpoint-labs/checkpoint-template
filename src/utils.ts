import { shortString } from 'starknet';

export function longStringToText(array: string[]): string {
  return array.reduce((acc, slice) => acc + shortString.decodeShortString(slice), '');
}
