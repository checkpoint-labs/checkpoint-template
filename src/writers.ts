import { CheckpointWriters } from '@snapshot-labs/checkpoint';
import { hexStrArrToStr, toAddress } from './utils';

export const writers: CheckpointWriters = {
  handleDeploy: async () => {
    // Run logic as at the time Contract was deployed.
  },

  handleNewPost: async ({ receipt, block, mysql }) => {
    const event = receipt.events[0] as any;
    const author = toAddress(event.data[0]);
    let content = '';
    let tag = '';
    const contentLength = BigInt(event.data[1]);
    const tagLength = BigInt(event.data[2 + Number(contentLength)]);
    const timestamp = (block as any).timestamp;
    const blockNumber = (block as any).block_number;

    // parse content bytes
    try {
      content = hexStrArrToStr(event.data, 2, contentLength);
    } catch (e) {
      console.error(`failed to decode content on block [${blockNumber}]: ${e}`);
      return;
    }

    // parse tag bytes
    try {
      tag = hexStrArrToStr(event.data, 3 + Number(contentLength), tagLength);
    } catch (e) {
      console.error(`failed to decode tag on block [${blockNumber}]: ${e}`);
      return;
    }

    // post object matches fields of Post type in schema.gql
    const post = {
      id: `${author}/${receipt.transaction_hash}`,
      author,
      content,
      tag,
      tx_hash: receipt.transaction_hash,
      created_at: timestamp,
      created_at_block: blockNumber
    };

    // table names are `lowercase(TypeName)s` and can be interacted with sql
    await mysql.queryAsync('INSERT IGNORE INTO posts SET ?', [post]);
  }
};
