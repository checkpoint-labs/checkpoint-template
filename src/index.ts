import 'dotenv/config';
import express from 'express';
import cors from 'cors';
import path from 'path';
import fs from 'fs';
import Checkpoint, { LogLevel, evm } from '@snapshot-labs/checkpoint';
import config from './config.json';
import * as writers from './writers';
import ProxyFactory from './abis/ProxyFactory.json';

const dir = __dirname.endsWith('dist/src') ? '../' : '';
const schemaFile = path.join(__dirname, `${dir}../src/schema.gql`);
const schema = fs.readFileSync(schemaFile, 'utf8');

const indexer = new evm.EvmIndexer(writers);

// Initialize checkpoint
const checkpoint = new Checkpoint(config, indexer, schema, {
  logLevel: LogLevel.Info,
  resetOnConfigChange: true,
  abis: {
    ProxyFactory
  }
  // prettifyLogs: true, // uncomment in local dev
});

// resets the entities already created in the database
// ensures data is always fresh on each re-run
checkpoint.reset().then(() => {
  // start the indexer
  checkpoint.start();
});

const app = express();
app.use(express.json({ limit: '4mb' }));
app.use(express.urlencoded({ limit: '4mb', extended: false }));
app.use(cors({ maxAge: 86400 }));

// mount Checkpoint's GraphQL API on path /
app.use('/', checkpoint.graphql);

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => console.log(`Listening at http://localhost:${PORT}`));
