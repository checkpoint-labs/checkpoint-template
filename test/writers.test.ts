import { mockDeep } from 'jest-mock-extended';
import { AsyncMySqlPool, ContractSourceConfig } from '@snapshot-labs/checkpoint';
import { Transaction, TransactionReceipt, GetBlockResponse } from 'starknet';
import { handleNewPost } from '../src/writers';

describe('Writers', () => {
  describe('handleNewPost', () => {
    it('should create posts correctly', async () => {
      const block = mockDeep<GetBlockResponse>({
        timestamp: 1652650963200,
        block_number: 500
      });
      const tx = mockDeep<Transaction>();
      const source = mockDeep<ContractSourceConfig>();
      const receipt = {
        events: [
          {
            from_address: '0x521f830ce263a6f0969f914d34e7001bc953c65ff04f316a9d2923eae145967',
            keys: ['0x8fee3afaa1e95194c589c2c7ff049f7667a0ddad86dd1e5172aa5d2053ca88'],
            data: [
              '0x0',
              '0x2',
              '0x5468697320697320612072616e646f6d20706f737420776974682061207261',
              '0x6e646f6d206e756d6265723a20393531',
              '0x1',
              '0x70726976617465'
            ]
          }
        ],
        transaction_hash: '0x5fa53789e3fc35c9dd40914e664d8508023289bdf986ebe30a540698fbeb026'
      };
      const mockMysql = mockDeep<AsyncMySqlPool>();

      await handleNewPost({
        tx,
        source,
        block,
        receipt: receipt as unknown as TransactionReceipt,
        mysql: mockMysql
      });

      expect(mockMysql.queryAsync.mock.calls).toMatchSnapshot();
    });
  });
});
