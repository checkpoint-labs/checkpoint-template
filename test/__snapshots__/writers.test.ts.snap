// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Writers handleNewPost should create posts correctly 1`] = `
Array [
  Array [
    "INSERT IGNORE INTO posts SET ?",
    Array [
      Object {
        "author": "0x0",
        "content": "This is a random post with a random number: 951",
        "created_at": 1652650963200,
        "created_at_block": 500,
        "id": "0x0/0x5fa53789e3fc35c9dd40914e664d8508023289bdf986ebe30a540698fbeb026",
        "tag": "private",
        "tx_hash": "0x5fa53789e3fc35c9dd40914e664d8508023289bdf986ebe30a540698fbeb026",
      },
    ],
  ],
]
`;
